src/app/CommandSender.h

```diff
@@ -36,15 +36,69 @@
 #include <system/SystemPacketBuffer.h>
 
 #include <app/Command.h>
+#include <app/MessageDef/CommandPath.h>
+#include <app/MessageDef/StatusElement.h>
 
 #define COMMON_STATUS_SUCCESS 0
 
 namespace chip {
 namespace app {
 
-class CommandSender : public Command, public Messaging::ExchangeDelegate
+class CommandSender final : public Command, public Messaging::ExchangeDelegate
 {
 public:
+    // TODO: Move various CommandResponse related callbacks to this delegate.
+    class Delegate
+    {
+    public:
+        virtual ~Delegate() = default;
+
+        /**
+         * OnResponse will be called when the response from server arrived.
+         *  - When StatusElement is received, aStatus will be filled with the content of the StatusElement, aData will be a
+         * uninitialized TLVReader.
+         *  - When command specific response is received, aStatus will be filled with success status, aData will be initialized with
+         * a struct of command specify response.
+         *
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction, MUST NOT be freed in this
+         * callback.
+         * @param[in] aPath: The command path field in invoke command response.
+         * @param[in] aStatus: The status element, filled with actual StatusElement when a StatusElement is received, when command
+         * specific response is received, aStatus will always be filled with success status code.
+         * @param[in] aData: The command data, or a uninitialized TLVReader if a StatusElement is received.
+         */
+        virtual void OnResponse(const CommandSender * apCommandSender, const CommandPath::Type & aPath,
+                                const StatusElement::Type & aStatus, TLV::TLVReader & aData)
+        {}
+
+        /**
+         * OnError will be called when error occurred when receiving / handling response from server:
+         * - CHIP_ERROR_TIMEOUT: The CommandSender failed to receive a expected response.
+         * - TLV Related erros: The CommandSender received a malformed response from server.
+         *
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction, MUST NOT be freed in this
+         * callback.
+         * @param[in] aError: The error occurred during the invoke command interaction.
+         */
+        virtual void OnError(const CommandSender * apCommandSender, CHIP_ERROR aError) {}
```

Comment:
We should be conveying the IM status codes as well as cluster-specific status codes (if one was present) through this API.

Please file a follow-up issue to fix this.